streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -730,16 +827,150 @@ private static void buildAssignedActiveTaskAndPartitionsList(final List<TaskId>
+            }
+
+            final Iterator<String> consumerIt = unfilledConsumers.iterator();
+            // Loop through the unfilled consumers and distribute tasks
+            while (consumerIt.hasNext()) {
+                final String consumer = consumerIt.next();
+                final List<TaskId> consumerAssignment = assignments.get(consumer);
+                final TaskId task = newTasks.poll();
+                if (task == null) {
+                    break;
+                }
+
+                consumerAssignment.add(task);
+                if (consumerAssignment.size() == maxTasksPerClient) {
+                    consumerIt.remove();
+                }
+            }
+        }
+
+        return assignments;
+    }
+
+    /**
+     * Get the previous consumer for the partitions of a task
+     *
+     * @param taskPartitions the TopicPartitions for a single given task
+     * @param clientOwnedPartitions the partitions owned by all consumers in a client
+     * @param allOwnedPartitions all partitions claimed as owned by any consumer in any client
+     * @return set of consumer(s) that previously owned the partitions in this task
+     *         empty set signals that it is a new task, or its previous owner is no longer in the group
+     */
+    Set<String> previousConsumersOfTaskPartitions(final Set<TopicPartition> taskPartitions,
+                                                 final Map<TopicPartition, String> clientOwnedPartitions,
+                                                 final Set<TopicPartition> allOwnedPartitions) {
+        // this "consumer" indicates a partition was owned by someone from another client -- we don't really care who
+        final String foreignConsumer = "";
+        final Set<String> previousConsumers = new HashSet<>();
+
+        for (final TopicPartition tp : taskPartitions) {
+            final String currentPartitionConsumer = clientOwnedPartitions.get(tp);
+            if (currentPartitionConsumer != null) {
+                previousConsumers.add(currentPartitionConsumer);
+            } else if (allOwnedPartitions.contains(tp)) {
+                previousConsumers.add(foreignConsumer);
+            }
+        }
+
+        return previousConsumers;
+    }
+
     // visible for testing
-    static List<List<TaskId>> interleaveTasksByGroupId(final Collection<TaskId> taskIds, final int numberThreads) {
+    static Map<String, List<TaskId>> interleaveTasksByGroupId(final Collection<TaskId> taskIds, final Set<String> consumers) {
         final LinkedList<TaskId> sortedTasks = new LinkedList<>(taskIds);
         Collections.sort(sortedTasks);
-        final List<List<TaskId>> taskIdsForConsumerAssignment = new ArrayList<>(numberThreads);
-        for (int i = 0; i < numberThreads; i++) {
-            taskIdsForConsumerAssignment.add(new ArrayList<>());
+        final Map<String, List<TaskId>> taskIdsForConsumerAssignment = new TreeMap<>();
```

Comment:
The motivation is to avoid the random order of consumers in HashMap, so that we hopefully end up with a similar task -> consumer assignment in subsequent rebalances.